File: otel-ios-mobile/Sources/OTelMobileSDK/Network/NetworkAvailabilityWatcher.swift
import Foundation

/// Listener invoked by `NetworkAvailabilityWatcher` on a genuine network transition.
/// Mirrors the Android `NetworkAvailabilityWatcher.Listener` functional interface.
public protocol NetworkAvailabilityListener: AnyObject {
    func onTransition(_ transition: NetworkAvailabilityWatcher.Transition)
}

/// Tracks network availability transitions and notifies listeners only on
/// **genuine** LOST → AVAILABLE transitions.
///
/// Why: `NWPathMonitor.pathUpdateHandler` fires for many non-transition
/// reasons (Wi-Fi handoff, DNS resolution, monitor start). A naive "every
/// path update → flush" wiring would trigger flushes constantly. This
/// watcher applies a small state machine so only the offline→online edge
/// produces a `.restored` event.
///
/// `NWPathMonitor` plumbing (the actual path-update bridge) is intentionally
/// **not** in this class — it arrives in a follow-up so the state-machine
/// logic stays unit-testable on the Foundation-only toolchain (no Network
/// framework dependency in tests).
///
/// See: docs/epics/NETWORK_RESTORED_FLUSH_EPIC.md (NF-008/NF-009).
public final class NetworkAvailabilityWatcher: @unchecked Sendable {

    /// Discrete network states the watcher tracks.
    private enum State {
        case unknown, lost, available
    }

    /// Transitions surfaced to listeners. Only `.restored` today; reserved
    /// enum for future events (e.g. `.degraded`, `.expensive`).
    public enum Transition: Sendable, Equatable {
        case restored
    }

    private let lock = NSLock()
    private var state: State = .unknown
    private var listeners: [WeakBox] = []

    public init() {}

    public func addListener(_ listener: NetworkAvailabilityListener) {
        lock.lock(); defer { lock.unlock() }
        // Prune any tombstoned weak refs while we hold the lock.
        listeners.removeAll { $0.value == nil }
        listeners.append(WeakBox(listener))
    }

    public func removeListener(_ listener: NetworkAvailabilityListener) {
        lock.lock(); defer { lock.unlock() }
        listeners.removeAll { $0.value === listener || $0.value == nil }
    }

    /// Called by the network-system adapter (NWPathMonitor) when a path
    /// becomes satisfied. Emits `.restored` **only** if the previous state
    /// was `.lost`.
    public func onAvailable() {
        let shouldNotify: Bool = {
            lock.lock(); defer { lock.unlock() }
            let wasLost = (state == .lost)
            state = .available
            return wasLost
        }()
        if shouldNotify {
            notifyAll(.restored)
        }
    }

    /// Called by the network-system adapter when the path becomes unsatisfied.
    public func onLost() {
        lock.lock(); defer { lock.unlock() }
        state = .lost
    }

    private func notifyAll(_ transition: Transition) {
        // Snapshot listeners under lock, fire callbacks outside it to avoid
        // re-entrant deadlocks if a listener calls back into the watcher.
        let snapshot: [NetworkAvailabilityListener] = {
            lock.lock(); defer { lock.unlock() }
            return listeners.compactMap { $0.value }
        }()
        for listener in snapshot {
            listener.onTransition(transition)
        }
    }

    /// Holds a weak ref so listeners that go out of scope are auto-collected.
    private final class WeakBox {
        weak var value: NetworkAvailabilityListener?
        init(_ value: NetworkAvailabilityListener) { self.value = value }
    }
}

File: otel-android-mobile/src/main/java/io/opentelemetry/android/mobile/network/NetworkAvailabilityWatcher.kt
/*
 * Copyright 2025 Barry Solomon
 * SPDX-License-Identifier: Apache-2.0
 */

package io.opentelemetry.android.mobile.network

import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicReference

/**
 * Tracks network availability transitions and notifies listeners only on
 * **genuine** LOST → AVAILABLE transitions.
 *
 * Why: `ConnectivityManager.NetworkCallback.onAvailable()` fires for many
 * non-transition reasons (Wi-Fi handoff, network validation, callback
 * registration). A naive "every onAvailable → flush" wiring would trigger
 * flushes constantly. This watcher applies a small state machine so that
 * only the offline→online edge produces a [Transition.Restored] event.
 *
 * Network-system wiring (the actual `ConnectivityManager.NetworkCallback`
 * adapter) is intentionally **not** in this class — it arrives in NF-002
 * so the state-machine logic can be unit-tested without Robolectric.
 *
 * See: `docs/epics/NETWORK_RESTORED_FLUSH_EPIC.md` (NF-001).
 */
class NetworkAvailabilityWatcher {

    /** Discrete network states the watcher tracks. */
    private enum class State { UNKNOWN, LOST, AVAILABLE }

    /** Transitions surfaced to listeners. Only [Restored] today; reserved enum for future events. */
    enum class Transition { Restored }

    /** Listener interface for transition events. */
    fun interface Listener {
        fun onTransition(transition: Transition)
    }

    private val state = AtomicReference(State.UNKNOWN)
    private val listeners = CopyOnWriteArrayList<Listener>()

    fun addListener(listener: Listener) {
        listeners.add(listener)
    }

    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }

    /**
     * Called by the network-system adapter when the OS reports a network is available.
     * Emits [Transition.Restored] **only** if the previous state was [State.LOST].
     */
    fun onAvailable() {
        val previous = state.getAndSet(State.AVAILABLE)
        if (previous == State.LOST) {
            notifyListeners(Transition.Restored)
        }
    }

    /** Called by the network-system adapter when the OS reports the network is lost. */
    fun onLost() {
        state.set(State.LOST)
    }

    private fun notifyListeners(transition: Transition) {
        listeners.forEach { it.onTransition(transition) }
    }
}
